File: .cph/Apna College/Sorting/QuickSort.cpp
 
#include<iostream>
using namespace std;
 
int partitian(int a[],int l,int r){
    int i=l-1;
    int pivot=a[r];

    for(int j=l;j<r;j++){
        if (a[j]<pivot)
        {
            i++;
            if(a[i]!=a[j]){
                int temp=a[i];
                a[i]=a[j];
                a[j]=temp;
            }
        }
    }
    a[r]=a[i+1];
    a[i+1]=pivot;
    return i+1;
}

void quicksort(int a[],int l,int r){
    if(l>=r) return;
    int pi = partitian(a,l,r);
    quicksort(a,l,pi-1);
    quicksort(a,pi+1,r);
}

int main()
{
    int n;
    cin>>n;
    int arr[n];
    for (int i = 0; i < n; i++)
    {
        cin>>arr[i];
    }
    quicksort(arr,0,n-1);
    for (int i = 0; i < n; i++)
    {
        cout<<arr[i]<<" ";
    }
    return 0;
}

File: Basic Algorithms/quick_sort.py
#time complexity n log n 
def partition(arr, left,right):
    pivot=arr[left]
    i=left
    j=right
    while i<j:
        while i<=right and arr[i]<=pivot:i+=1
        while j>=left and arr[j]>pivot:j-=1
        if i<j:arr[i],arr[j]=arr[j],arr[i]

    arr[left],arr[j]=arr[j],arr[left]
    return j

def quicksort(arr,left,right):
    if left<right:
        partition_index=partition(arr,left,right)
        quicksort(arr,left,partition_index-1)
        quicksort(arr,partition_index+1,right)

    
arr=[4,1,3,9,7]
print(arr)
quicksort(arr,0,4)
print(arr)